### PR TITLE
Improve thread safety around contextMap

### PR DESCRIPTION
**Pull request by watou**
Trying to find the proper fix for #2842 and #1419.  Found lack of thread safety around protected `contextMap` member of `AbstractGenericBindingProvider`.  Also found bindings performing unnecessary and bug-inducing changes in their overrides of `processBindingConfiguration` calls.

Please review this code carefully (as I'm sure you would anyway).  I've added line comments in the changes to explain my thinking.  These changes are not tested, but here to spur others' thinking about the origins of #2842 and #1419.

### CODE REVIEW

File: bundles/binding/org.openhab.binding.autelis/src/main/java/org/openhab/binding/autelis/internal/AutelisGenericBindingProvider.java [hunk 1]
```diff
@@ -80,13 +76,8 @@ public void processBindingConfiguration(String context, Item item, String bindin
...
-		
-		Set<Item> items = contextMap.get(context);
-		if (items == null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
-		}
-		items.add(item);
+
+		super.processBindingConfiguration(context, item, bindingConfig);
 	}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.autelis/src/main/java/org/openhab/binding/autelis/internal/AutelisGenericBindingProvider.java]: Removed code duplicated what super.processBindingConfiguration had already done, so moved super call to end.

File: bundles/binding/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMGenericBindingProvider.java [hunk 2]
```diff
@@ -124,6 +124,6 @@ public void removeConfigurations(String context) {
 
 	@Override
 	public Set<Item> getItemNamesByContext(String context) {
-		return new HashMap<String, Set<Item>>(super.contextMap).get(context);
+		return contextMap.get(context);
 	}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMGenericBindingProvider.java]: Replaced creation of duplicate contextMap and extraction of set from it, with simple extraction of set from contextMap.

File: bundles/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/DSCAlarmGenericBindingProvider.java [hunk 3]
```diff
@@ -107,13 +106,8 @@ public void processBindingConfiguration(String context, Item item, String bindin
...
 		addBindingConfig(item, config);
-		Set<Item> items = contextMap.get(context);
-		if (items == null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
-		}
-		items.add(item);
 		
+		super.processBindingConfiguration(context, item, bindingConfig);
 		logger.debug("processBindingConfiguration(): Item added: {} - {}",item.getName(),bindingConfig);
```
Comment (watou) [inline bundles/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/DSCAlarmGenericBindingProvider.java]: Removed code duplicated what super.processBindingConfiguration had already done, so moved super call to end.

File: bundles/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/config/EnoceanGenericBindingProvider.java [hunk 4]
```diff
@@ -77,17 +77,6 @@ protected void addBindingConfig(Item item, BindingConfig config) {
...
-        if (configuredItems != null) {
-            for (Item item : configuredItems) {
-                items.remove(item.getName());
-            }
-        }
-        super.removeConfigurations(context);
-    }
-
-    @Override
     public EnoceanParameterAddress getParameterAddress(String itemName) {
```
Comment (watou) [inline bundles/binding/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/config/EnoceanGenericBindingProvider.java]: This override performed an unnecessary one-by-one removal of items from the contextMap, but introduced problem where listeners would not be called and items would not be removed from bindingConfigs map.  Removing method altogether should remove the bug.

File: bundles/binding/org.openhab.binding.freeswitch/src/main/java/org/openhab/binding/freeswitch/internal/FreeswitchGenericBindingProvider.java [hunk 5]
```diff
@@ -74,13 +73,7 @@ public void processBindingConfiguration(String context, Item item, String bindin
...
 		addBindingConfig(item, config);
-		
-		Set<Item> items = contextMap.get(context);
-		if (items == null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
-		}
-		items.add(item);
+		super.processBindingConfiguration(context, item, bindingConfig);
 	}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.freeswitch/src/main/java/org/openhab/binding/freeswitch/internal/FreeswitchGenericBindingProvider.java]: Removed code duplicated what super.processBindingConfiguration had already done, so moved super call to end.

File: bundles/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/internal/HarmonyHubGenericBindingProvider.java [hunk 6]
```diff
@@ -105,12 +104,7 @@ public void processBindingConfiguration(String context, Item item, String bindin
...
 
 			addBindingConfig(item, config);		
-			Set<Item> items = contextMap.get(context);
-			if (items == null) {
-				items = new HashSet<Item>();
-				contextMap.put(context, items);
-			}
-			items.add(item);
+			super.processBindingConfiguration(context, item, bindingConfig);
 		} else {
```
Comment (watou) [inline bundles/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/internal/HarmonyHubGenericBindingProvider.java]: Removed code duplicated what super.processBindingConfiguration had already done, so moved super call to end.

File: bundles/binding/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/OmniLinkGenericBindingProvider.java [hunk 7]
```diff
@@ -75,12 +75,7 @@ public void processBindingConfiguration(String context, Item item,
...
 		addBindingConfig(item, config);
 
-		Set<Item> items = contextMap.get(context);
-		if (items == null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
-		}
-		items.add(item);
+		super.processBindingConfiguration(context, item, bindingConfig);
 	}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.omnilink/src/main/java/org/openhab/binding/omnilink/internal/OmniLinkGenericBindingProvider.java]: super.processBindingConfiguration was never called, but method attempted to achieve equivalent result by hand.

File: bundles/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/SatelGenericBindingProvider.java [hunk 8]
```diff
@@ -53,11 +53,11 @@ public void validateItemType(Item item, String bindingConfig) throws BindingConf
 	public void processBindingConfiguration(String context, Item item, String bindingConfig)
 			throws BindingConfigParseException {
 		logger.trace("Processing binding configuration for item {}", item.getName());
-		super.processBindingConfiguration(context, item, bindingConfig);
 
 		SatelBindingConfig bc = this.createBindingConfig(bindingConfig);
 		logger.trace("Adding binding configuration for item {}: {}", item.getName(), bc);
 		addBindingConfig(item, bc);
+		super.processBindingConfiguration(context, item, bindingConfig);
 	}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/SatelGenericBindingProvider.java]: Moved super.processBindingConfiguration to end of override, since previous steps could fail and we don't want invalid binding config objects in the system that were rejected.

File: bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireGenericBindingProvider.java [hunk 9]
```diff
@@ -61,12 +61,11 @@ public void validateItemType(Item pvItem, String pvBindingConfig) throws Binding
...
 	@Override
 	public void processBindingConfiguration(String pvContext, Item pvItem, String pvBindingConfig) throws BindingConfigParseException {
-		super.processBindingConfiguration(pvContext, pvItem, pvBindingConfig);
-
 		OneWireBindingConfig pvDevicePropertyBindingConfig = OneWireBindingConfigFactory.createOneWireDeviceProperty(pvItem, pvBindingConfig);
 
 		addBindingConfig(pvItem, pvDevicePropertyBindingConfig);
 
+		super.processBindingConfiguration(pvContext, pvItem, pvBindingConfig);
 	}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireGenericBindingProvider.java]: Moved super.processBindingConfiguration to end of override, since previous steps could fail and we don't want invalid binding config objects in the system that were rejected.

File: bundles/binding/org.openhab.binding.xpl/src/main/java/org/openhab/binding/xpl/internal/XplGenericBindingProvider.java [hunk 10]
```diff
@@ -159,6 +158,7 @@ else if (configParts[1].equalsIgnoreCase("COMMAND"))
 	    } 
 		
 		addBindingConfig(item, config);
+		super.processBindingConfiguration(context, item, bindingConfig);
 	}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.xpl/src/main/java/org/openhab/binding/xpl/internal/XplGenericBindingProvider.java]: Moved super.processBindingConfiguration to end of override, since previous steps could fail and we don't want invalid binding config objects in the system that were rejected.

File: bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java [hunk 11]
```diff
@@ -42,7 +41,7 @@
 	private Set<BindingChangeListener> listeners = new CopyOnWriteArraySet<BindingChangeListener>();
 
 	/** caches binding configurations. maps itemNames to {@link BindingConfig}s */
-	protected Map<String, BindingConfig> bindingConfigs = new ConcurrentHashMap<String, BindingConfig>(new WeakHashMap<String, BindingConfig>());
+	protected Map<String, BindingConfig> bindingConfigs = new ConcurrentHashMap<String, BindingConfig>();
```
Comment (watou) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Constructor was only copying, not wrapping, the new WeakHashMap, so removed.

File: bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java [hunk 12]
```diff
@@ -77,27 +76,31 @@ public void processBindingConfiguration(String context, Item item, String bindin
...
-			contextMap.put(context, items);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if (items==null) {
+				items = new HashSet<Item>();
+				contextMap.put(context, items);
+			}
+
+			items.add(item);
 		}
```
Comment (watou) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Needed to make contextMap.get and contextMap.put one atomic operation, so put in a synchronized block.

File: bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java [hunk 13]
```diff
@@ -77,27 +76,31 @@ public void processBindingConfiguration(String context, Item item, String bindin
...
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if(items!=null) {
+				for(Item item : items) {
+					// we remove all binding configurations for all items
+					bindingConfigs.remove(item.getName());
+					notifyListeners(item);
+				}
+				contextMap.remove(context);
 			}
```
Comment (watou) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Needed to make contextMap.get and contextMap.remove one atomic operation, so put in a synchronized block.

**Comment by watou** [comment]
This PR goes with [this Google Group thread](https://groups.google.com/forum/#!category-topic/openhab/p46JDUy3Ev8).

File: bundles/binding/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMGenericBindingProvider.java [hunk 14]
```diff
@@ -121,9 +119,4 @@ public void removeConfigurations(String context) {
 		}
 		return deviceConsumptionItems;
 	}
-
-	@Override
-	public Set<Item> getItemNamesByContext(String context) {
-		return new HashMap<String, Set<Item>>(super.contextMap).get(context);
-	}
-}
```
Comment (watou) [inline bundles/binding/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMGenericBindingProvider.java]: Removed unused `getItemNamesByContext` method since it exposes non-threadsafe Set of Items.

File: bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java (hunk 13 above)
Comment (mrguessed) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: You have an internal/framework lock held when you make calls, via `notifyListeners`, to non-framework code.

It would be better if you bucketed _what_ to callout, while inside the framework lock, and then made the callouts themselves outside the lock...otherwise, depending upon the locking model of all of the called entities (Bindings), you could create a deadlock.


PS: You may want to run the Eclipse code-formatting function over the changes to align with the std openHAB uses.

Comment (watou) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Thanks very much for having a look, @mrguessed.  I was hoping you would take an interest.  I will move the loop to call `notifyListeners` outside the lock as you suggest.  I didn't want to re-format the changes for fear it would hurt readability, but I see your point.  Any other thoughts/ideas/warnings most appreciated.

File: bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java [hunk 15]
```diff
@@ -77,27 +76,34 @@ public void processBindingConfiguration(String context, Item item, String bindin
...
+		}
+		if (items != null) {
+			for (Item item : items) {
 				// we remove all binding configurations for all items
 				bindingConfigs.remove(item.getName());
 				notifyListeners(item);
 			}
-			contextMap.remove(context);
 		}
 	}
```
Comment (watou) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Is this what you meant, @mrguessed ?

Comment (mrguessed) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Generally, yes.  That said, I'm not sure how much cross-consistency is expected, by Core, between the `contextMap` and the `bindingConfig` entries.  

The change now has us removing the `bindingConfig` entry _after_ the `contextMap` entry, which reverses the logical ordering from the original code.  I'm not familiar enough with this layer to know if that's a problem or not.

Comment (watou) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Thanks for taking another look.  I also don't know if there are side effects in changing the ordering, but what I do know is that now, if multiple threads are calling `removeConfigurations` with the same `context`, as I suspect is what is causing to the NPEs, only the first will have any effect, and the others will essentially be no-ops.  Unless there is an intervening `processBindingConfiguration`, in which case the next thread to call `removeConfigurations` will be iterating over a completely different `Set` from the previous thread to enter `removeConfigurations`.  In other words, I don't see how multiple threads can step on each other when contending for the `contextMap` after these changes.

There are most likely higher-level coordination issues when files are reloaded, but I believe (and tell me if I'm wrong) that at least this code, in isolation, appears threadsafe.

Comment (mrguessed) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: It's threadsafe, but my concern about remove [contextMap] before notify [bindingConfig] needs to be validated.  For example, with the changes in place someone previously using the `contextMap` object, via some sort of call into these Core methods during the `notifyListeners` phase might get a surprise ;)  

I don't know if that can actually happen, since I haven't looked over this part in that much detail, so it's theoretical at this point.

Comment (watou) [inline bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java]: Some bindings' BindingProvider implementations have a `getItem` method that references the `contextMap` to find an item by name, but I don't know if this code is ever reached from `notifyListeners` in Core.  My opinion, subject to @teichsta or @kaikreuzer saying otherwise, is that my changes are more correct, in that during a `removeConfigurations` call, bindings ought not be trying to locate items by name, since the bindings are in the process of being torn down.

**Comment by watou** [comment]
It appears that the only reason current master works when *.items files are touched is _because of_ a thread concurrency issue in higher-level code.  Looking at the following sequence with the Ecobee binding, for example, the current Ecobee binding ought to create a NullPointerException every time:

* touch ecobee.items -> ... ->
 *  org.openhab.model.item.internal.GenericItemProvider.processBindingConfigsFromModel ->
  * org.openhab.model.item.binding.AbstractGenericBindingProvider.removeConfigurations ->
   * for each item in the contextMap set:
     * remove the item by name from bindingConfigs, THEN
     * notifyListeners, which calls bindingChanged in the EcobeeBinding, which then consults bindingConfigs for the item name, but it was just removed, so NPE.

So the only way this works is if EcobeeBinding.bindingChanged is never called, or some other thread has already added back the refreshed binding config object in a race condition.

In other words, while this PR fixes thread concurrency issues in AbstractGenericBindingProvider, there are higher-level side effects that actually depend on thread concurrency side effects for the current bindings to work.  I don't know to what degree these issues exist in OH2, but it's very seriously worth a look if there is no intention to fix these issues in OH1.

**Comment by teichsta** [comment]
i'll have a look into this in the course of the next week. Thanks so much for all the analysis you've so far!

**Comment by watou** [comment]
Hey @teichsta, I squashed this commit just now.  Please have a look when you can.  Thanks!

**Comment by teichsta** [comment]
i had a look into the code and i agree with @watous changes. For the next weeks (after the merge) we should keep an eye on all issues that somehow "smell" like caused by this issue.

**Comment by teichsta** [comment]
Thanks a million @watou and @mrguessed for driving this issue and the fruitful discussions around this PR!

**Comment by watou** [comment]
@teichsta Thanks, Thomas.  Did you mean to close this PR without merge? :smile:

**Comment by teichsta** [comment]
(thanks :blush:)

**Comment by lolodomo** [comment]
Comparing 1.7.1 and current 1.8 snapshot, I can confirm that something was really fixed as the refresh of my items is now working well. I still see the Null Pointer Exception but apparently it does not break the reload process anymore.
Good.

**Comment by watou** [comment]
Thank you for the report, @lolodomo.  Do you have an example of the NullPointerException stack trace against the 1.8 snapshot?  I'm really interested in seeing what it looks like, as I was expecting that there would be no exceptions thrown after the change.

**Comment by lolodomo** [comment]
Here is the stack trace I get:

12:15:44.906 [ERROR] [.o.m.c.i.folder.FolderObserver:105  ] - An unexpected exception has occured
java.lang.NullPointerException: null
        at org.openhab.model.item.binding.AbstractGenericBindingProvider.removeConfigurations(AbstractGenericBindingProvider.java:104) ~[na:na]
        at org.openhab.model.item.internal.GenericItemProvider.processBindingConfigsFromModel(GenericItemProvider.java:164) ~[na:na]
        at org.openhab.model.item.internal.GenericItemProvider.modelChanged(GenericItemProvider.java:390) ~[na:na]
        at org.openhab.model.core.internal.ModelRepositoryImpl.notifyListeners(ModelRepositoryImpl.java:159) ~[na:na]
        at org.openhab.model.core.internal.ModelRepositoryImpl.addOrRefreshModel(ModelRepositoryImpl.java:100) ~[na:na]
        at org.openhab.model.core.internal.folder.FolderObserver.checkFolder(FolderObserver.java:142) ~[na:na]
        at org.openhab.model.core.internal.folder.FolderObserver.run(FolderObserver.java:99) ~[na:na]

**Comment by watou** [comment]
Thanks for the stack trace. I will investigate this shortly.

**Comment by watou** [comment]
Hi @lolodomo, it's very odd that either `bindingConfigs` or `item` would be `null` at [line 104 in AbstractGenericBindingProvider.java](https://github.com/openhab/openhab/blob/master/bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java#L104).  I am running a private build from today's master branch, installed from .deb files on an RPi2/wheezy, including these addons:
```
org.openhab.action.ecobee-1.8.0-SNAPSHOT.jar
org.openhab.action.mail-1.8.0-SNAPSHOT.jar
org.openhab.action.twitter-1.8.0-SNAPSHOT.jar
org.openhab.binding.astro-1.8.0-SNAPSHOT.jar
org.openhab.binding.ecobee-1.8.0-SNAPSHOT.jar
org.openhab.binding.http-1.8.0-SNAPSHOT.jar
org.openhab.binding.mqtt-1.8.0-SNAPSHOT.jar
org.openhab.binding.mqttitude-1.8.0-SNAPSHOT.jar
org.openhab.binding.nest-1.8.0-SNAPSHOT.jar
org.openhab.binding.networkhealth-1.8.0-SNAPSHOT.jar
-rw-r--r-- 1 openhab openhab   10185 Oct 31 13:40 org.openhab.binding.ntp-1.8.0-SNAPSHOT.jar
-rw-r--r-- 1 openhab openhab 1618995 Oct 31 13:41 org.openhab.binding.xbmc-1.8.0-SNAPSHOT.jar
-rw-r--r-- 1 openhab openhab  613092 Oct 31 13:44 org.openhab.io.myopenhab-1.8.0-SNAPSHOT.jar
-rw-r--r-- 1 openhab openhab  832963 Oct 31 13:44 org.openhab.persistence.mysql-1.8.0-SNAPSHOT.jar
```
I have 5 .items files and when I `touch` each one, I don't get any exceptions.  So....
* are you using a recent snapshot build of runtime and all addons?
* which bindings are you running?
* which hardware?

**Comment by lolodomo** [comment]
Hi @watou
I am running a 1.8 snapshot downloaded the 28th of October.

Here are my installed bindings (downloaded the same day):
org.openhab.action.mios-1.8.0-SNAPSHOT.jar
org.openhab.binding.astro-1.8.0-SNAPSHOT.jar
org.openhab.binding.freebox-1.8.0-SNAPSHOT.jar
org.openhab.binding.http-1.8.0-SNAPSHOT.jar
org.openhab.binding.hue-1.8.0-SNAPSHOT.jar
org.openhab.binding.mios-1.8.0-SNAPSHOT.jar
org.openhab.binding.netatmo-1.8.0-SNAPSHOT.jar
org.openhab.binding.ntp-1.8.0-SNAPSHOT.jar
org.openhab.binding.powermax-1.7.0-SNAPSHOT.jar
org.openhab.binding.rfxcom-1.8.0-SNAPSHOT.jar
org.openhab.binding.weather-1.8.0-SNAPSHOT.jar
org.openhab.persistence.logging-1.8.0-SNAPSHOT.jar
org.openhab.persistence.rrd4j-1.8.0-SNAPSHOT.jar

I have only 1 items file.
My hardware is a RaspBerry Pi.

**Comment by watou** [comment]
Thank you, @lolodomo.  I see that before my code change, [what was line 97](https://github.com/openhab/openhab/blob/6fbe28aecaee6f56195e57d2b814a616aa27236c/bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java#L97) is the same as what is [now line 104](https://github.com/openhab/openhab/blob/master/bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java#L104), so the same code is failing:
```java
for (Item item : items) {
	// we remove all binding configurations for all items
	bindingConfigs.remove(item.getName());  // this line throws NullPointerException
	notifyListeners(item);
}
```
...except now I've made the modification to `contextMap` atomic and threadsafe before the NPE, and this seems to have improved matters for you.  So the NPE means either a `null` key was added to the `bindingConfigs` map, or the `bindingConfigs` map itself is `null`.  This seems to be a timing issue that appears on the slower single-core Raspberry Pi 1 but not on the much faster quad-core Raspberry Pi 2.  Interesting.  So this PR set out to fix handling of the `contextMap` and did, but another race condition seems to be in play.

**Comment by lolodomo** [comment]
@watou: I will try to run OH (same bindings) with my Windows 7 PC to confirm that the NPE is not happening on a fast hardware.

**Comment by watou** [comment]
@lolodomo: That would be interesting to learn whether that makes the difference.  The slower RPi 1 seems to show timing issues in some cases that faster systems don't show.  Ideally we could identify and fix them so system performance/number of cores makes no difference other than time.
